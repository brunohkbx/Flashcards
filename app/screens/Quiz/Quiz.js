import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { View, ScrollView } from 'react-native';
import { Button, DefaultTheme } from 'react-native-paper';
import Container from '../../components/Container';
import FlashcardCollapsed from '../../components/FlashcardCollapsed';
import { Fadable } from '../../components/Animations';
import Result from './Result';

const ButtonContainer = styled(View)`
  flex: 1;
  flex-direction: row;
  justify-content: space-around;
  align-items: flex-end;
  padding-vertical: 30;
`;

export class Quiz extends Component {
  state = {
    currentQuestion: 1,
    correct: 0,
    incorrect: 0,
    fade: false
  }

  scorePositive = () => this.setState(prevState => {
    if (this.state.fade)
      return;

    return { correct: prevState.correct + 1, fade: true };
  })

  scoreNegative = () => this.setState(prevState => {
    if (this.state.fade)
      return;

    return { incorrect: prevState.incorrect + 1, fade: true };
  })

  renderNextFlashcard = () => this.setState(prevState => {
    return { currentQuestion: prevState.currentQuestion + 1, fade: false };
  })

  handleRestartQuiz = () => this.setState(
    { currentQuestion: 1, correct: 0, incorrect: 0, fade: false }
  )

  totalQuestions = (() => { return this.props.deck.questions.length; })()

  isFinished = () => {
    const { currentQuestion } = this.state;

    return (currentQuestion - 1) === this.totalQuestions;
  }

  render() {
    const { deck, settings } = this.props;
    const { currentQuestion, fade, correct } = this.state;

    return (
      <Container>
        {!this.isFinished() &&
          <ScrollView contentContainerStyle={{flexGrow: 1}}>
            <Fadable fade={fade} onAnimationEnd={this.renderNextFlashcard}>
              <FlashcardCollapsed
                currentQuestion={currentQuestion}
                totalQuestions={this.totalQuestions}
                flashcard={deck.questions[currentQuestion - 1]}
              />
            </Fadable>
            <ButtonContainer>
              <Button
                mode="outlined"
                color={DefaultTheme.colors.accent}
                icon="close"
                onPress={this.scoreNegative}
              >
                I was wrong
              </Button>
              <Button
                mode="contained"
                primary
                icon="done"
                onPress={this.scorePositive}
              >
                I was right
              </Button>
            </ButtonContainer>
          </ScrollView>
        }
        {this.isFinished() &&
          <Result
            score={correct}
            totalQuestions={this.totalQuestions}
            onRestartQuiz={this.handleRestartQuiz}
            receiveNotifications={settings.receiveNotifications}
          />
        }
      </Container>
    );
  }
}

const mapStateToProps = ({ decks, settings }, { navigation }) => {
  const { deckId } = navigation.state.params;

  return {
    deck: decks[deckId],
    settings
  };
};

export default connect(mapStateToProps)(Quiz);
