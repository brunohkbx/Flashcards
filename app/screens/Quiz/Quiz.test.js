import React from 'react';
import { shallow } from 'enzyme';
import { Quiz } from './Quiz';

describe('Quiz', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      deck: {
        title: 'React',
        questions: [
          {
            question: 'What is React?',
            answer: 'A library for managing user interfaces'
          }
        ]
      },
      settings: {
        receiveNotifications: false
      }
    }, propOverrides);

    const wrapper = shallow(<Quiz {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    return { wrapper, wrapperInstance };
  };

  it('renders properly when the quiz is not over yet', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders properly when the quiz has ended', () => {
    const { wrapper } = setup();

    wrapper.setState({ currentQuestion: 2, correct: 1 });

    expect(wrapper).toMatchSnapshot();
  });

  describe('scorePositive', () => {
    it('returns undefined when component is fading', () => {
      const { wrapperInstance } = setup();

      wrapperInstance.setState({ fade: true});

      expect(wrapperInstance.scorePositive()).toBeUndefined();
    });

    it('increments the number of correct answer and fade the element', () => {
      const { wrapper, wrapperInstance } = setup();
      const prevState = wrapper.state();

      wrapperInstance.scorePositive();

      expect(wrapper.state()).toMatchObject(
        { correct: prevState.correct + 1, fade: true }
      );
    });
  });

  describe('scoreNegative', () => {
    it('returns undefined when component is fading', () => {
      const { wrapperInstance } = setup();

      wrapperInstance.setState({ fade: true});

      expect(wrapperInstance.scoreNegative()).toBeUndefined();
    });

    it('increments the number of incorrect answer and fade the element', () => {
      const { wrapper, wrapperInstance } = setup();
      const prevState = wrapper.state();

      wrapperInstance.scoreNegative();

      expect(wrapper.state()).toMatchObject(
        { incorrect: prevState.incorrect + 1, fade: true }
      );
    });
  });

  describe('renderNextFlashcard', () => {
    it('increments the currentQuestion and unfade the element', () => {
      const { wrapper, wrapperInstance } = setup();
      const prevState = wrapper.state();

      wrapperInstance.renderNextFlashcard();

      expect(wrapper.state()).toMatchObject(
        { currentQuestion: prevState.currentQuestion + 1, fade: false }
      );
    });
  });

  describe('handleRestartQuiz', () => {
    it('resets the state', () => {
      const { wrapper, wrapperInstance } = setup();

      wrapperInstance.handleRestartQuiz();

      expect(wrapper.state()).toMatchObject(
        { currentQuestion: 1, correct: 0, incorrect: 0, fade: false }
      );
    });
  });

  describe('totalQuestions', () => {
    it('returns the amount of questions from the deck', () => {
      const { wrapperInstance } = setup();
      const totalQuestions = wrapperInstance.props.deck.questions.length;

      expect(wrapperInstance.totalQuestions).toEqual(totalQuestions);
    });
  });

  describe('isFinished', () => {
    it('checks if currentQuestion - 1 is equal to totalQuestions', () => {
      const { wrapperInstance } = setup();
      wrapperInstance.setState({ currentQuestion: 4 });
      wrapperInstance.totalQuestions = 3;

      expect(wrapperInstance.isFinished()).toBeTruthy();
    });
  });
});
