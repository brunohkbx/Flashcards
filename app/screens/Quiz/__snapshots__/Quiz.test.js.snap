// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz renders properly when the quiz has ended 1`] = `
<Styled(Component)>
  <Result
    onRestartQuiz={[Function]}
    receiveNotifications={false}
    score={1}
    totalQuestions={1}
  />
</Styled(Component)>
`;

exports[`Quiz renders properly when the quiz is not over yet 1`] = `
<Styled(Component)>
  <ScrollViewMock
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <Fadable
      fade={false}
      onAnimationEnd={[Function]}
    >
      <FlashcardCollapsed
        currentQuestion={1}
        flashcard={
          Object {
            "answer": "A library for managing user interfaces",
            "question": "What is React?",
          }
        }
        totalQuestions={1}
      />
    </Fadable>
    <Styled(Component)>
      <withTheme(Button)
        color="#03dac4"
        icon="close"
        mode="outlined"
        onPress={[Function]}
      >
        I was wrong
      </withTheme(Button)>
      <withTheme(Button)
        icon="done"
        mode="contained"
        onPress={[Function]}
        primary={true}
      >
        I was right
      </withTheme(Button)>
    </Styled(Component)>
  </ScrollViewMock>
</Styled(Component)>
`;
